man/oftpd.8
.TH OFTPD 8 "2001-04-03" Linux "Linux System Administration"
.SH NAME 
oftpd \- anonymous, read-only FTP server
.SH SYNOPSIS
.B oftpd [-p|--port
.I port
.B ] [-r|--pasv-range
.I low high
.B ] [-i|--interface
.I ip-address
.B ] [-m|--max-clients
.I num
.B ] [-l|--local
.I local-logging
.B ] [-N|--nodetach]
.I user-name root-directory
.SH DESCRIPTION
.B oftpd
is an anonymous-only FTP server.  

Although it must be started by the root user, after initialization it
runs as the user specified by 
.I user-name
parameter.  This should be a user with minimal permissions, preferably
from an account set up for this purpose only.

The server uses
.BR chroot (2)
to change the root directory of the server to the 
.I root-directory
directory.  When a user connects, this is the directory that they will
start in, and is the top of their directory tree.

When FTP clients connect, they may log in as "ftp" or "anonymous".
Typically they will then send their e-mail address as password (most web
browsers send a browser identifier rather than the user's e-mail
address).  Standard Unix file permissions are enforced for downloading,
meaning clients may download any file 
.I user-name
has permission to read.  No uploads are permitted.  All client activity
is logged (see 
.B DIAGNOSTICS
below).
.SH OPTIONS
.IP "-p|--port port"
Use the specified
.I port 
to listen for client connections.  If not specified, the default FTP
port (number 21) is used.
.IP "-r|--pasv-range low high"
Limit the range for ports used in passive mode from
.I low
to
.I high
 .  If not specified 1024 and 65535 are used.
.IP "-i|--interface ip-address"
Use the interface connected to the IP address
.I ip-address
to accept connections.  If not specified, the server listens on all
interfaces.
.IP "-m|--max-clients num"
Accept at most 
.I num
simultaneous clients.  If not specified, 250 will be the limit.
.IP "-l|--local local-logging"
Normally oftpd logs messages to syslog as the FTP daemon.  With this
option, the specified 
.I local-logging
level will be used instead.  Valid numbers are 0-8.
.IP "-N|--nodetach"
Do not run in the background; for running from
.BR init (8)
or for testing.
.SH DIAGNOSTICS
Syntax errors will result in a help message being displayed, and a
non-zero exit code returned.  Otherwise the server will exit without
output, and return zero.

After the server has started, you should check the appropriate log
produced by 
.BR sysklogd (8)
for the FTP daemon to insure that startup completed correctly.  Certain
errors occur only after the server has disconnected from the TTY, so
can only be recorded via the log mechanism.

If the server must terminate for any reason, it will also be logged, as
will other non-fatal internal errors.  They are logged with attention to
how serious the condition is believed to be, as documented in the
.BR syslog (3)
library call.

Client activity will also be logged through this mechanism.  This
includes connect and disconnect (or rejection due to too many
simultaneous users), the e-mail address reported as password, and file
transfers.  All client commands are logged as sent at the 
.B DEBUG 
level.  To fully monitor client activity you may configure
.BR sysklogd (8)
to record these.
.SH NOTES
.B oftpd
does not use 
.BR inetd (8)
to run.  It is a stand-alone server.  There is no need to configure the 
.I /etc/inetd.conf
file to run it.  In fact, there should be no entry for FTP there at all.

You can use the
.BR kill (1)
command to stop
.B oftpd.
When the server receives SIGHUP or SIGINT, it will stop listening for
new FTP connections, and a new FTP server may be started.  Any existing
connections to the old server will continue to function normally until
the client disconnects or times out.  After all clients connections have
closed, the server will exit.  

To shutdown the server and close all client connections immediately, use
SIGKILL.
.SH AUTHOR
Shane Kerr <shane@time-travellers.org>
.SH "SEE ALSO"
.BR ftp (1)



NEWS
2003-08-13 <wk@gnupg.org>

  New option -r, --pasv-range to limit the range for passive ports to
  specific firewall rules.

  Print a warning if the /dev directory does not exists below the
  chroot.  It is required so that a syslogd restart won't harm.
  Ignore the /dev directory in the ftp session.

  Fixed some minor problems.

2000-12-13

  IPv6 code is maintained by Mauro Tortonesi <mauro@ferrara.linux.it>.
  You can download the latest patches for oftpd from the Project6
  website:

    http://project6.ferrara.linux.it

  Thanks Mauro!


src/oftpd.c
/*
 * $Id$
 */

#include <config.h>
#include <stdio.h>
#include <unistd.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <fcntl.h>
#include <string.h>
#include <errno.h>
#include <signal.h>
#include <pwd.h>
#include <syslog.h>
#include <pthread.h>
#include <stdlib.h>
#include <sys/socket.h>
#include <sys/un.h>

#include "oftpd.h"
#include "ftp_listener.h"
#include "error.h"

/* put our executable name here where everybody can see it */
static const char *exe_name = "oftpd";

int pasv_port_low = 1024;
int pasv_port_high = MAX_PORT;

/* The debug flags.  */
int debug_flags;

static void daemonize();
static void print_usage(const char *error);

int main(int argc, char *argv[])
{
    int i;
    
    long num;
    char *endptr;
    int port;
    int max_clients;

    char *user_ptr;
    char *dir_ptr;
    char *address;
    
    char temp_buf[256];
    char errbuf[ERRBUF_SIZE];

    struct passwd *user_info;
    error_t err;

    ftp_listener_t ftp_listener;

    int detach;

    int log_facility;

    sigset_t term_signal;
    int sig;

    /* grab our executable name */
    if (argc > 0) {
        exe_name = argv[0];
    }

    /* verify we're running as root */
    if (geteuid() != 0) {
        if ( argc > 1 && !strcmp (argv[1], "--version")) {
                printf ("%s %s\n", PACKAGE_NAME, PACKAGE_VERSION);
                exit (0);
        }
        fprintf(stderr, "%s: program needs root permission to run\n", exe_name);
        exit(1);

    }

    /* default command-line arguments */
    port = FTP_PORT;
    user_ptr = NULL;
    dir_ptr = NULL;
    address = FTP_ADDRESS;
    max_clients = MAX_CLIENTS;
    detach = 1;
    log_facility = LOG_FTP;

    /* check our command-line arguments */
    /* we're stubbornly refusing to use getopt(), because we can */
    /* :) */ 
    for (i=1; i<argc; i++) {
        
        /* flags/optional arguments */
        if (argv[i][0] == '-') {
            if (strcmp(argv[i], "-p") == 0 
                || strcmp(argv[i], "--port") == 0) {
                if (++i >= argc) {
                    print_usage("missing port number");
                    exit(1);
                }
                num = strtol(argv[i], &endptr, 0);
                if ((num < MIN_PORT) || (num > MAX_PORT) || (*endptr != '\0')) {

                    snprintf(temp_buf, sizeof(temp_buf), 
                             "port must be a number between %d and %d",
                             MIN_PORT, MAX_PORT);
                    print_usage(temp_buf);

                    exit(1);
                }
                port = num;
            }
            else if (strcmp(argv[i], "-r") == 0 
                     || strcmp(argv[i], "--pasv-range") == 0) {
                i += 2;  
                if (i >= argc) {
                    print_usage("invalid passive port range");
                    exit(1);
                }
                num = strtol(argv[i-1], &endptr, 0);
                if ((num < 1024) || (num > MAX_PORT) || (*endptr != '\0')) {

                    snprintf(temp_buf, sizeof(temp_buf), 
                             "low passive port must be a number between 1024 and %d",
                             MAX_PORT);
                    print_usage(temp_buf);

                    exit(1);
                }
                pasv_port_low = num;
                num = strtol(argv[i], &endptr, 0);
                if ((num < pasv_port_low) || (num > MAX_PORT) || (*endptr != '\0')) {

                    snprintf(temp_buf, sizeof(temp_buf), 
                             "high passive port must be a number between %d and %d",
                             pasv_port_low, MAX_PORT);
                    print_usage(temp_buf);

                    exit(1);
                }
                pasv_port_high = num;
            } else if (strcmp(argv[i], "-h") == 0
                       || strcmp(argv[i], "--help") == 0) {
                print_usage(NULL);
                exit(0);
            } else if (strcmp(argv[i], "-i") == 0
                       || strcmp(argv[i], "--interface") == 0) {
                if (++i >= argc) {
                    print_usage("missing interface");
                    exit(1);
                }
                address = argv[i];
            } else if (strcmp(argv[i], "-m") == 0
                       || strcmp(argv[i], "--max-clients") == 0) {
                if (++i >= argc) {
                    print_usage("missing number of max clients");
                    exit(1);
                }
                num = strtol(argv[i], &endptr, 0);
                if ((num < MIN_NUM_CLIENTS) || (num > MAX_NUM_CLIENTS) 
                    || (*endptr != '\0')) {

                    snprintf(temp_buf, sizeof(temp_buf),
                        "max clients must be a number between %d and %d",
                        MIN_NUM_CLIENTS, MAX_NUM_CLIENTS);
                    print_usage(temp_buf);

                    exit(1);
                }
                max_clients = num;
            } else if (strcmp(argv[i], "-N") == 0 
                       || strcmp(argv[i], "--nodetach") == 0) {
                detach = 0;
            } else if (strcmp(argv[i], "-d") == 0
                       || strcmp(argv[i], "--debug") == 0) { 
	        /* We just add another flag for each -d option.  */
	        debug_flags = (debug_flags << 1) | 1;
            } else if (strcmp(argv[i], "-l") == 0
                       || strcmp(argv[i], "--local") == 0) {
                if (++i >= argc) {
                    print_usage("missing number for local facility logging");
                    exit(1);
                }
                switch (argv[i][0]) {
                    case '0': 
                        log_facility = LOG_LOCAL0;
                        break;
                    case '1': 
                        log_facility = LOG_LOCAL1;
                        break;
                    case '2': 
                        log_facility = LOG_LOCAL2;
                        break;
                    case '3': 
                        log_facility = LOG_LOCAL3;
                        break;
                    case '4': 
                        log_facility = LOG_LOCAL4;
                        break;
                    case '5': 
                        log_facility = LOG_LOCAL5;
                        break;
                    case '6': 
                        log_facility = LOG_LOCAL6;
                        break;
                    case '7': 
                        log_facility = LOG_LOCAL7;
                        break;
                }
            } else if (strcmp(argv[i], "--version") == 0) {
                printf ("%s %s\n", PACKAGE_NAME, PACKAGE_VERSION);
                exit (0);
            } else {
                print_usage("unknown option");
                exit(1);
            }

        /* required parameters */
        } else {
            if (user_ptr == NULL) {
                user_ptr = argv[i];
            } else if (dir_ptr == NULL) {
                dir_ptr = argv[i];
            } else {
                print_usage("too many arguments on the command line");
                exit(1);
            }
        }
    }
    if ((user_ptr == NULL) || (dir_ptr == NULL)) {
        print_usage("missing user and/or directory name");
        exit(1);
    }

    user_info = getpwnam(user_ptr);
    if (user_info == NULL) {
        fprintf(stderr, "%s: invalid user name\n", exe_name);
        exit(1);
    }

    /* become a daemon */
    if (detach) {
        daemonize();
    }

    /* avoid SIGPIPE on socket activity */
    signal(SIGPIPE, SIG_IGN);         

    /* log the start time */
    openlog(NULL, LOG_NDELAY, log_facility);
    syslog(LOG_INFO,"Starting, version %s, as PID %d",
           PACKAGE_VERSION, getpid());

    /* change to root directory */
    if (chroot(dir_ptr) != 0) {
        syslog(LOG_ERR, "error with root directory; %s\n", 
          strerror_r(errno, errbuf, ERRBUF_SIZE));
        exit(1);
    }
    if (chdir("/") != 0) {
        syslog(LOG_ERR, "error changing directory; %s\n",
               strerror_r(errno, errbuf, ERRBUF_SIZE));
        exit(1);
    }

    /* Check that a /dev directory exists, so that syslog works
       properly even after a syslogd restart. */
    {
      struct stat stat_buf;

      if (stat( "/dev", &stat_buf)) {
        syslog (LOG_ERR, "required `%s/dev' directory is missing: %s\n",
                dir_ptr, strerror_r (errno, errbuf, ERRBUF_SIZE));
      }
#ifndef STATS_MACRO_BROKEN
      if (!S_ISDIR(stat_buf.st_mode)) {
#else
      if (S_ISDIR(stat_buf.st_mode)) {
#endif
          syslog (LOG_ERR, "`%s/dev' is not a directory\n", dir_ptr);
      }
    }


    /* create our main listener */
    if (!ftp_listener_init(&ftp_listener, 
                           address,
                           port,
                           max_clients,
                           INACTIVITY_TIMEOUT, 
                           &err)) 
    {
        syslog(LOG_ERR, "error initializing FTP listener; %s",
          error_get_desc(&err));
        exit(1);
    }

    /* set user to be as inoffensive as possible */
    if (setgid(user_info->pw_gid) != 0) {
        syslog(LOG_ERR, "error changing group; %s",
               strerror_r(errno, errbuf, ERRBUF_SIZE));
        exit(1);
    }
    if (setuid(user_info->pw_uid) != 0) {
        syslog(LOG_ERR, "error changing group; %s",
               strerror_r(errno, errbuf, ERRBUF_SIZE));
        exit(1);
    }

    /* start our listener */
    if (ftp_listener_start(&ftp_listener, &err) == 0) {
        syslog(LOG_ERR, "error starting FTP service; %s", error_get_desc(&err));
        exit(1);
    }

    /* wait for a SIGTERM and exit gracefully */
    sigemptyset(&term_signal);
    sigaddset(&term_signal, SIGTERM);
    sigaddset(&term_signal, SIGINT);
    pthread_sigmask(SIG_BLOCK, &term_signal, NULL);
    sigwait(&term_signal, &sig);
    if (sig == SIGTERM) {
        syslog(LOG_INFO, "SIGTERM received, shutting down");
    } else { 
        syslog(LOG_INFO, "SIGINT received, shutting down");
    }
    ftp_listener_stop(&ftp_listener);
    syslog(LOG_INFO, "all connections finished, FTP server exiting");
    exit(0);
}

static void print_usage(const char *error)
{
    if (error != NULL) {
        fprintf(stderr, "%s: %s\n", exe_name, error);
    }
    fprintf(stderr, 
           " Syntax: %s [ options... ] user_name root_directory\n", exe_name);
    fprintf(stderr, 
           " Options:\n"
           " -p, --port <num>\n"
           "     Set the port to listen on (Default: %d)\n"
           " -i, --interface <IP Address>\n"
           "     Set the interface to listen on (Default: all)\n"
           " -m, --max-clients <num>\n"
           "     Set the number of clients allowed at one time (Default: %d)\n"
           "-l, --local <local-logging>\n"
           "     Use LOCAL facility for syslog, local-logging is 0 to 7\n"
           " -N, --nodetach\n"
           "     Do not detach from TTY and become a daemon\n",
           DEFAULT_FTP_PORT, MAX_CLIENTS);
}

static void daemonize()
{
    int fork_ret;
    int max_fd;
    int null_fd;
    int fd;
    char errbuf[ERRBUF_SIZE];

    null_fd = open("/dev/null", O_RDWR);
    if (null_fd == -1) {
        fprintf(stderr, "%s: error opening null output device; %s\n", exe_name, 
          strerror_r(errno, errbuf, ERRBUF_SIZE));
        exit(1);
    }

    max_fd = sysconf(_SC_OPEN_MAX);
    if (max_fd == -1) {
        fprintf(stderr, "%s: error getting maximum open file; %s\n", exe_name, 
          strerror_r(errno, errbuf, ERRBUF_SIZE));
        exit(1);
    }


    fork_ret = fork();
    if (fork_ret == -1) {
        fprintf(stderr, "%s: error forking; %s\n",
                exe_name, strerror_r(errno, errbuf, ERRBUF_SIZE));
        exit(1);
    }
    if (fork_ret != 0) {
        exit(0);
    }
    if (setsid() == -1) {
        fprintf(stderr, "%s: error creating process group; %s\n", exe_name, 
          strerror_r(errno, errbuf, ERRBUF_SIZE));
        exit(1);
    }
    fork_ret = fork();
    if (fork_ret == -1) {
        fprintf(stderr, "%s: error forking; %s\n",
                exe_name, strerror_r(errno, errbuf, ERRBUF_SIZE));
        exit(1);
    }
    if (fork_ret != 0) {
        exit(0);
    }
    if (dup2(null_fd, 0) == -1) {
        syslog(LOG_ERR, "error setting input to null; %s", 
          strerror_r(errno, errbuf, ERRBUF_SIZE));
        exit(1);
    }
    if (dup2(null_fd, 1) == -1) {
        syslog(LOG_ERR, "error setting output to null; %s", 
          strerror_r(errno, errbuf, ERRBUF_SIZE));
        exit(1);
    }
    if (dup2(null_fd, 2) == -1) {
        syslog(LOG_ERR, "error setting error output to null; %s", 
          strerror_r(errno, errbuf, ERRBUF_SIZE));
        exit(1);
    }
    for (fd=3; fd<max_fd; fd++) {
        close(fd);
    }
}

